### PR TITLE
Supplement the description of IPVS

### PR DESCRIPTION
**Pull request by RifeWang**
### Description

Supplement IPVS scheduling algorithms that added in https://github.com/kubernetes/kubernetes/issues/114619

### Issue

Closes: #

### CODE REVIEW

**Comment by sftim** [comment]
LGTM for docs

**Comment by sftim** [comment]
@kubernetes/sig-network-pr-reviews is this correct?

**Comment by RifeWang** [comment]
I tested `mh` and added explanations for the `mh` flag  `mh-port`.
image
image

source code: https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L1092

File: content/en/docs/reference/networking/virtual-ips.md [hunk 1]
```diff
@@ -249,6 +249,16 @@ these are:
   waiting for a fast one; if all servers are busy, the algorithm falls back to the `sed`
   behavior.
 
+* `mh` (Maglev Hashing): Assigns incoming jobs based on Google's Maglev hashing algorithm,
```
Comment (tengqm) [inline content/en/docs/reference/networking/virtual-ips.md:252]: Better provide a link to the algorithm.

File: content/en/docs/reference/networking/virtual-ips.md [hunk 2]
```diff
@@ -249,6 +249,16 @@ these are:
...
+  providing an almost equal share of jobs to each real server and provides minimal disruption.
```
Comment (tengqm) [inline content/en/docs/reference/networking/virtual-ips.md:253]: providing ... and provides ....

Syntax error?

File: content/en/docs/reference/networking/virtual-ips.md [hunk 3]
```diff
@@ -249,6 +249,16 @@ these are:
...
+  When the set of real servers changes, a connection will likely be sent to the same real server
+  as it was before. This scheduler has two flags: `mh-fallback`, which enables fallback to a different
+  server if the selected server was unavailable, and `mh-port`, which adds the source port number to
```
Comment (tengqm) [inline content/en/docs/reference/networking/virtual-ips.md:256]:
```suggestion
  server if the selected server is unavailable, and `mh-port`, which adds the source port number to
```

File: content/en/docs/reference/networking/virtual-ips.md [hunk 4]
```diff
@@ -249,6 +249,16 @@ these are:
...
+  the hash computation. We are using `mh-port` by default.
```
Comment (tengqm) [inline content/en/docs/reference/networking/virtual-ips.md:257]: Who are "we"?

File: content/en/docs/reference/networking/virtual-ips.md [hunk 5]
```diff
@@ -249,6 +249,16 @@ these are:
...
+
+These scheduling algorithms are configured through the
+[`ipvs.scheduler`](/docs/reference/config-api/kube-proxy-config.v1alpha1/#kubeproxy-config-k8s-io-v1alpha1-KubeProxyIPVSConfiguration) field in the kube-proxy configuration.
```
Comment (tengqm) [inline content/en/docs/reference/networking/virtual-ips.md:260]:
```suggestion
[`ipvs.scheduler`](/docs/reference/config-api/kube-proxy-config.v1alpha1/#kubeproxy-config-k8s-io-v1alpha1-KubeProxyIPVSConfiguration)
field in the kube-proxy configuration.
```

**Review by shaneutt — COMMENTED** [review]
/cc @aojea @danwinship

**Comment by aojea** [comment]
/assign @uablrek 

Lars is the resident expert in IPVS

**Comment by uablrek** [comment]
I am a bit reluctant to describe maglev-hashing (mh) like this. While accurate, it gives a false impression that `mh` will actually improve disruption when configured in kube-proxy/ipvs. It will not. The reason is that mh is intended for use in a stateless load-balancer, while `kube-proxy` is stateful (using conntracking).

The "set of real servers" is endpoints in K8s (PODs most often). And a "change" in that set is (most often) a change of "replicas" in a Deployment. In a scale-out all connections are preserved (because kube-proxy keeps a state). So we have "no disruption" on scale-out with default setting of kube-proxy in any mode. That can't be improved.

So, I would like to avoid arguments to keep proxy-mode=ipvs to "minimize disruption with mh". I am more inclined to let proxy-mode=ipvs silently fade into the sunset in favor of proxy-mode=nftables.

A possible use-case I have investigated is when an external router re-distributes traffic to nodes for some reason. For instance all external traffic goes to "worker1", and then suddenly is re-routed to "worker2". In that case _**all**_ external connections are lost in K8s. Now, _in theory_, with "mh" and a permissive conntrack setting (meaning that tcp packets not belonging to a connection are allowed), packet arriving to "worker2" would be routed to the same real-server (PODs) as before (when traffic went to "worker1"). However, _I have never got that working_. (have anyone else?)

**Comment by uablrek** [comment]
There is more to read in https://github.com/kubernetes/kubernetes/issues/114619 about this.

And for the record, I think the maglev hashing algorithm is _absolutely brilliant_! So my intention is not to diminish it in any way.

**Comment by RifeWang** [comment]
> @RifeWang Would you like revise the explanation of the 'mh' mode so that it is more objective and accurate?

Or do not add the description for `mh`, only keep the instructions for configuring the scheduling algorithm field

**Comment by uablrek** [comment]
Users that want to use "mh" have probably read something very similar to your description. So perhaps a warning:

> In K8s using "mh" will not improve disruption. It will work same as source-hashing (sh), but with ports

**Comment by RifeWang** [comment]
* `mh` (Maglev Hashing): Assigns incoming jobs based on [Google's Maglev hashing algorithm](https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/44824.pdf),
  providing an almost equal share of jobs to each real server and minimizing disruption.
  When the set of real servers changes, a connection will likely be sent to the same real server
  as before. This scheduler has two flags: `mh-fallback`, which enables fallback to a different
  server if the selected server is unavailable, and `mh-port`, which adds the source port number to
  the hash computation. When using `mh`, IPVS uses the `mh-port` flag by default.

{{< warning >}}
In proxy-mode=ipvs using "mh" will not improve disruption. It will work same as source-hashing (sh),
but with ports.
{{< /warning >}}

**Comment by RifeWang** [comment]
* `mh` (Maglev Hashing): Assigns incoming jobs based on [Google's Maglev hashing algorithm](https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/44824.pdf),
  providing an almost equal share of jobs to each real server ~~and minimizing disruption~~.
  ~~When the set of real servers changes, a connection will likely be sent to the same real server
  as before.~~ This scheduler has two flags: `mh-fallback`, which enables fallback to a different
  server if the selected server is unavailable, and `mh-port`, which adds the source port number to
  the hash computation. When using `mh`, IPVS always uses the `mh-port` flag ~~by default~~, meaning that
  `mh` may not improve disruption as much as expected.

**Comment by uablrek** [comment]
2 is incorrect. The use of ports in hashing does not affect disruption (not in kube-proxy, nor in maglev). So in that respect 1 is better.

However IMHO 1 still gives the impression that "mh" improves disruption. The statements about disruption are true for a stateless loadbalancer, but not for kube-proxy, which is stateful and using NAT (even with "mh"). Can you just skip it?

```
* `mh` (Maglev Hashing): Assigns incoming jobs based on [Google's Maglev hashing algorithm](https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/44824.pdf),
  This scheduler has two flags: `mh-fallback`, which enables fallback to a different
  server if the selected server is unavailable, and `mh-port`, which adds the source port number to
  the hash computation. When using `mh`, IPVS uses the `mh-port` flag by default.
  In proxy-mode=ipvs "mh" will work as source-hashing (sh), but with ports.
```

The "sh" scheduler can be configured to use ports in ipvs, but to do so in `kube-proxy` breaks backward compatibility. A K8s user may use "sh" to make sure all connects from a source ends up in the same endpoint (kind of a stateless affinity). But some users may not like "rr" and want a distribution similar to proxy-mode=nftables, and "mh" provides that.

**Comment by RifeWang** [comment]
In fact, IPVS always uses the `mh-port` flag and does not provide users with any configuration fields for this flag. 
Therefore,  `IPVS always uses the mh-port flag`  better than `IPVS uses the mh-port flag by default` ?

**Comment by RifeWang** [comment]
PR updated

**Review by uablrek — APPROVED** [review]

File: content/en/docs/reference/networking/virtual-ips.md [hunk 6]
```diff
@@ -249,6 +249,17 @@ these are:
   waiting for a fast one; if all servers are busy, the algorithm falls back to the `sed`
   behavior.
 
+* `mh` (Maglev Hashing): Assigns incoming jobs based on
+  [Google's Maglev hashing algorithm](https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/44824.pdf),
+  This scheduler has two flags: `mh-fallback`, which enables fallback to a different
+  server if the selected server is unavailable, and `mh-port`, which adds the source port number to
+  the hash computation. When using `mh`, IPVS always uses the `mh-port` flag.
```
Comment (uablrek) [inline content/en/docs/reference/networking/virtual-ips.md:256]: I missed this: "IPVS always uses" here should rather be "kube-proxy always sets". And perhaps mention that mh-fallback isn't used

**Comment by tengqm** [comment]
I think this is ready to go, once the additional nits are solved.
/approve

File: content/en/docs/reference/networking/virtual-ips.md (hunk 6 above)
Comment (RifeWang) [reply content/en/docs/reference/networking/virtual-ips.md:256]: When using `mh`, kube-proxy always sets the `mh-port` flag and does not enable the `mh-fallback` flag.

**Review by uablrek — APPROVED** [review]

**Comment by tengqm** [comment]
/hold cancel
Thanks.

**Pull request merged**